src/dfm/projects/production/model_training/scripts/continue_mistral.py

```diff
@@ -0,0 +1,275 @@
+                        warnings.warn("The dataset reached end and the iterator is reset to the start.")
+                    else:
+                        more_examples = False
+                        break
+            tokenized_inputs = self.tokenizer(buffer, truncation=False)["input_ids"]
+            all_token_ids = []
+            for tokenized_input in tokenized_inputs:
+                all_token_ids.extend(tokenized_input + [self.concat_token_id])
+            examples = []
+            for i in range(0, len(all_token_ids), self.seq_length):
+                input_ids = all_token_ids[i : i + self.seq_length]
+                if len(input_ids) == self.seq_length:
+                    examples.append(input_ids)
+            if self.shuffle:
+                random.shuffle(examples)
+            for example in examples:
+                self.current_size += 1
+                yield {
+                    "input_ids": torch.LongTensor(example),
+                    "labels": torch.LongTensor(example),
+                }
+
+
+def chars_token_ratio(dataset, tokenizer, nb_examples=400):
+    """
+    Estimate the average number of characters per token in the dataset.
+    """
+    total_characters, total_tokens = 0, 0
+    for _, example in tqdm(zip(range(nb_examples), iter(dataset)), total=nb_examples):
+        text = prepare_sample_text(example)
+        total_characters += len(text)
+        if tokenizer.is_fast:
+            total_tokens += len(tokenizer(text).tokens())
+        else:
+            total_tokens += len(tokenizer.tokenize(text))
+
+    return total_characters / total_tokens
+
+parser = HfArgumentParser(ScriptArguments)
+script_args = parser.parse_args_into_dataclasses()[0]
+
+# Step 1: Load the dataset
+tokenizer = AutoTokenizer.from_pretrained(script_args.model_name)
+dataset = load_dataset(script_args.dataset_name, split="train[:20000]")
+dataset = dataset.train_test_split(test_size=0.01)
+
+def prepare_dialogue(example):
+    text = ""
+    for idx, msg in enumerate(example["data"]):
+        if idx % 2 == 0:
+            text += f"<|user|>\n{msg}{tokenizer.eos_token}\n"
+        else:
+            text += f"<|assistant|>\n{msg}{tokenizer.eos_token}\n"
+    example["text"] = text
+    return example
+
+def tokenize(element):
+    outputs = tokenizer(
+        element["text"],
+        truncation=True,
+        max_length=script_args.seq_length,
+        return_overflowing_tokens=True,
+        return_length=True,
+    )
+    input_batch = []
+    for length, input_ids in zip(outputs["length"], outputs["input_ids"]):
+        if length == script_args.seq_length:
+            input_batch.append(input_ids)
+    return {"input_ids": input_batch}
+
+dataset = dataset.map(prepare_dialogue, num_proc=4, remove_columns=["id", "data"])
+#tokenized_datasets = dataset.map(
+#    tokenize, batched=True, remove_columns=dataset["train"].column_names, num_proc=4
+#)
+
+chars_per_token = 3.6 # TODO
```

Comment:
?

Reply:
This file is one of the unused ones btw, should probably just be removed.

Reply:
Ahh then let us just remove it